### PR TITLE
[urho3d] Add new port

### PR DESCRIPTION
**Pull request by koprok**
Add Urho3D port.

### CODE REVIEW

**Comment by NancyLi1013** [comment]
Hi @koprok 

Thanks for your PR.

The failures on Windows like this:

```
D:\buildtrees\urho3d\src\1.7.1-314e41264a.clean\Source\ThirdParty\SDL\src\events\SDL_events.c
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(2501): error C2118: negative subscript
```

The failures on osx like this:

```
/Users/vagrant/Data/buildtrees/urho3d/src/1.7.1-314e41264a.clean/Source/ThirdParty/Bullet/src/LinearMath/btVector3.h:343:7: error: argument value 10880 is outside the valid range [0, 255] [-Wargument-outside-range]
                y = bt_splat_ps(y, 0x80);
                    ^~~~~~~~~~~~~~~~~~~~
/Users/vagrant/Data/buildtrees/urho3d/src/1.7.1-314e41264a.clean/Source/ThirdParty/Bullet/src/LinearMath/btVector3.h:47:32: note: expanded from macro 'bt_splat_ps'
#define bt_splat_ps( _a, _i )  bt_pshufd_ps((_a), BT_SHUFFLE(_i,_i,_i,_i) )
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/vagrant/Data/buildtrees/urho3d/src/1.7.1-314e41264a.clean/Source/ThirdParty/Bullet/src/LinearMath/btVector3.h:45:35: note: expanded from macro 'bt_pshufd_ps'
#define bt_pshufd_ps( _a, _mask ) _mm_shuffle_ps((_a), (_a), (_mask) )
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include/xmmintrin.h:2596:11: note: expanded from macro '_mm_shuffle_ps'
  (__m128)__builtin_ia32_shufps((__v4sf)(__m128)(a), (__v4sf)(__m128)(b), \
          ^
1 error generated.
[466/1218] /Library/Developer/CommandLineTools/usr/bin/c++ -DURHO3D_ANGELSCRIPT -DURHO3D_CXX11 -DURHO3D_FILEWATCHER -DURHO3D_IK -DURHO3D_LOGGING -DURHO3D_LUA -DURHO3D_NAVIGATION -DURHO3D_NETWORK -DURHO3D_PHYSICS -DURHO3D_PROFILING -DURHO3D_STATIC_DEFINE -DURHO3D_THREADING -DURHO3D_URHO2D -DURHO3D_WEBP -I/Users/vagrant/Data/buildtrees/urho3d/src/1.7.1-314e41264a.clean/Source/ThirdParty/Bullet/src -fPIC -std=c++11 -Wno-invalid-offsetof -march=native -pthread -Qunused-arguments -fcolor-diagnostics -g -DDEBUG -D_DEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -MD -MT Source/ThirdParty/Bullet/CMakeFiles/Bullet.dir/src/BulletCollision/BroadphaseCollision/btCollisionAlgorithm.cpp.o -MF Source/ThirdParty/Bullet/CMakeFiles/Bullet.dir/src/BulletCollision/BroadphaseCollision/btCollisionAlgorithm.cpp.o.d -o Source/ThirdParty/Bullet/CMakeFiles/Bullet.dir/src/BulletCollision/BroadphaseCollision/btCollisionAlgorithm.cpp.o -c /Users/vagrant/Data/buildtrees/urho3d/src/1.7.1-314e41264a.clean/Source/ThirdParty/Bullet/src/BulletCollision/BroadphaseCollision/btCollisionAlgorithm.cpp
```

There are some thirdParty libraries included in the sources. Actually, we recommend to use these ports that provided in vcpkg if these are required dependencies.

So could you please try to update these parts?

**Comment by koprok** [comment]
Hello @NancyLi1013 

Thank you for your response. I will try to solve these errors. Unfortunately they don't happen on my own Windows PC here. Maybe it's because of different MSVC or Windows SDK versions.

About the 3rd party libraries: Unfortunately Urho3D contains forked versions of these libraries. I don't think it is possible to easily use the ports provided in vcpkg.

**Comment by NancyLi1013** [comment]
The failures on x64-windows:

```
FAILED: bin/DetourTileCache.dll Source/ThirdParty/DetourTileCache/DetourTileCache.lib 
cmd.exe /C "cd . && D:\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe -E vs_link_dll --intdir=Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x64\link.exe"  Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir\Source\DetourTileCache.cpp.obj Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir\Source\DetourTileCacheBuilder.cpp.obj  /out:bin\DetourTileCache.dll /implib:Source\ThirdParty\DetourTileCache\DetourTileCache.lib /pdb:bin\DetourTileCache.pdb /dll /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
LINK Pass 1: command "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x64\link.exe Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir\Source\DetourTileCache.cpp.obj Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir\Source\DetourTileCacheBuilder.cpp.obj /out:bin\DetourTileCache.dll /implib:Source\ThirdParty\DetourTileCache\DetourTileCache.lib /pdb:bin\DetourTileCache.pdb /dll /version:0.0 /machine:x64 /nologo /debug /INCREMENTAL kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir/intermediate.manifest Source\ThirdParty\DetourTileCache\CMakeFiles\DetourTileCache.dir/manifest.res" failed (exit code 1120) with the following output:
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "void * __cdecl dtAlloc(unsigned __int64,enum dtAllocHint)" (?dtAlloc@@YAPEAX_KW4dtAllocHint@@@Z) referenced in function "public: unsigned int __cdecl dtTileCache::init(struct dtTileCacheParams const *,struct dtTileCacheAlloc *,struct dtTileCacheCompressor *,struct dtTileCacheMeshProcess *)" (?init@dtTileCache@@QEAAIPEBUdtTileCacheParams@@PEAUdtTileCacheAlloc@@PEAUdtTileCacheCompressor@@PEAUdtTileCacheMeshProcess@@@Z)
DetourTileCacheBuilder.cpp.obj : error LNK2001: unresolved external symbol "void * __cdecl dtAlloc(unsigned __int64,enum dtAllocHint)" (?dtAlloc@@YAPEAX_KW4dtAllocHint@@@Z)
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "void __cdecl dtFree(void *)" (?dtFree@@YAXPEAX@Z) referenced in function "public: __cdecl dtTileCache::~dtTileCache(void)" (??1dtTileCache@@QEAA@XZ)
DetourTileCacheBuilder.cpp.obj : error LNK2001: unresolved external symbol "void __cdecl dtFree(void *)" (?dtFree@@YAXPEAX@Z)
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "bool __cdecl dtCreateNavMeshData(struct dtNavMeshCreateParams *,unsigned char * *,int *)" (?dtCreateNavMeshData@@YA_NPEAUdtNavMeshCreateParams@@PEAPEAEPEAH@Z) referenced in function "public: unsigned int __cdecl dtTileCache::buildNavMeshTile(unsigned int,class dtNavMesh *)" (?buildNavMeshTile@dtTileCache@@QEAAIIPEAVdtNavMesh@@@Z)
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "public: unsigned int __cdecl dtNavMesh::addTile(unsigned char *,int,int,unsigned int,unsigned int *)" (?addTile@dtNavMesh@@QEAAIPEAEHHIPEAI@Z) referenced in function "public: unsigned int __cdecl dtTileCache::buildNavMeshTile(unsigned int,class dtNavMesh *)" (?buildNavMeshTile@dtTileCache@@QEAAIIPEAVdtNavMesh@@@Z)
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "public: unsigned int __cdecl dtNavMesh::removeTile(unsigned int,unsigned char * *,int *)" (?removeTile@dtNavMesh@@QEAAIIPEAPEAEPEAH@Z) referenced in function "public: unsigned int __cdecl dtTileCache::buildNavMeshTile(unsigned int,class dtNavMesh *)" (?buildNavMeshTile@dtTileCache@@QEAAIIPEAVdtNavMesh@@@Z)
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "public: unsigned int __cdecl dtNavMesh::getTileRefAt(int,int,int)const " (?getTileRefAt@dtNavMesh@@QEBAIHHH@Z) referenced in function "public: unsigned int __cdecl dtTileCache::buildNavMeshTile(unsigned int,class dtNavMesh *)" (?buildNavMeshTile@dtTileCache@@QEAAIIPEAVdtNavMesh@@@Z)
DetourTileCache.cpp.obj : error LNK2019: unresolved external symbol "void (__cdecl*__cdecl dtAssertFailGetCustom(void))(char const *,char const *,int)" (?dtAssertFailGetCustom@@YAP6AXPEBD0H@ZXZ) referenced in function "public: unsigned int __cdecl dtTileCache::buildNavMeshTile(unsigned int,class dtNavMesh *)" (?buildNavMeshTile@dtTileCache@@QEAAIIPEAVdtNavMesh@@@Z)
DetourTileCacheBuilder.cpp.obj : error LNK2001: unresolved external symbol "void (__cdecl*__cdecl dtAssertFailGetCustom(void))(char const *,char const *,int)" (?dtAssertFailGetCustom@@YAP6AXPEBD0H@ZXZ)
bin\DetourTileCache.dll : fatal error LNK1120: 7 unresolved externals
```

**Comment by NancyLi1013** [comment]
This port failed on all triplets now.

Could you please look into the failures? @koprok 

Please see the detail https://dev.azure.com/vcpkg/public/_build/results?buildId=46044&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by NancyLi1013** [comment]
Hi @koprok 

Could you please merge mater to retrigger CI?

**Comment by koprok** [comment]
@NancyLi1013 , all builds have passed, except arm and uwp. I think these are not supported by urho3d. Can you now merge the pull request?

File: scripts/ci.baseline.txt [hunk 1]
```diff
@@ -1626,6 +1626,9 @@ unrar:x64-osx=fail
 unrar:x64-uwp=fail
 unrar:x64-windows-static=fail
 urdfdom:x64-windows-static=fail
+urho3d:arm64-windows=fail
+urho3d:arm-uwp=fail
+urho3d:x64-uwp=fail
```
Comment (NancyLi1013) [inline scripts/ci.baseline.txt:1631]: Please add `supports: !(arm | uwp)` to CONTROL file and remove these lines from ci.baseline.txt,

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: scripts/ci.baseline.txt (hunk 1 above)
Comment (koprok) [reply scripts/ci.baseline.txt]: @NancyLi1013 , done

**Comment by NancyLi1013** [comment]
```
Found the following errors:
Error: While reading versions for port urho3d from file: C:\a\1\s\versions\u-\urho3d.json
       File declares version `2021-03-01` with SHA: 71df9dc8a48be2a53bd1d6b079ba3799c50f418c
       But local port with the same verion has a different SHA: 75c562c3779b97bf4a0cc0d6259d4b528fe9af05
       Please update the port's version fields and then run:

           vcpkg x-add-version urho3d

       to add a new version.
```
Please run  `vcpkg x-add-version --overwrite-version urho3d` to solve the failures on x86-windows. @koprok

**Comment by koprok** [comment]
> ```
> Found the following errors:
> Error: While reading versions for port urho3d from file: C:\a\1\s\versions\u-\urho3d.json
>        File declares version `2021-03-01` with SHA: 71df9dc8a48be2a53bd1d6b079ba3799c50f418c
>        But local port with the same verion has a different SHA: 75c562c3779b97bf4a0cc0d6259d4b528fe9af05
>        Please update the port's version fields and then run:
> 
>            vcpkg x-add-version urho3d
> 
>        to add a new version.
> ```
> 
> Please run `vcpkg x-add-version --overwrite-version urho3d` to solve the failures on x86-windows. @koprok

@NancyLi1013, done. But now x64-linux build fails and I don't think the reason is in the port itself.

File: ports/urho3d/CONTROL [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Source: urho3d
+Version: 2021-03-01
+Homepage: https://github.com/urho3d/Urho3D
+Description: Urho3D is a free lightweight, cross-platform 2D and 3D game engine implemented in C++ and released under the MIT license. Greatly inspired by OGRE and Horde3D.
+Supports: !(arm|uwp)
```
Comment (NancyLi1013) [inline ports/urho3d/CONTROL:5]: Can you update this as vcpkg.json file? We prefer to use manifests for new ports. Please see https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/pr-review-checklist.md
Also, please also update `version `filed as `version_date`. Please see https://github.com/microsoft/vcpkg/blob/master/docs/users/versioning.md#version-schemes

File: ports/urho3d/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        shared_libs.patch
+)
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+    set(URHO3D_LIB_TYPE "STATIC")
+else()
+    set(URHO3D_LIB_TYPE "SHARED")
+endif()
+
+vcpkg_configure_cmake(
```
Comment (NancyLi1013) [inline ports/urho3d/portfile.cmake:19]: This function has been deprecated now. Can you update as the new one `vcpkg_cmake_configure()`? We prefer to use suggested functions for new ports.

File: ports/urho3d/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
```
Comment (NancyLi1013) [inline ports/urho3d/portfile.cmake:21]:
```suggestion
```
If we use the new function, this argument is no longer needed here. Please see https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/ports/vcpkg-cmake/vcpkg_cmake_configure.md.

File: ports/urho3d/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    OPTIONS
+        -DURHO3D_LIB_TYPE=${URHO3D_LIB_TYPE}
+        -DURHO3D_C++11=ON
+        -DURHO3D_PCH=OFF
+)
+
+vcpkg_install_cmake()
```
Comment (NancyLi1013) [inline ports/urho3d/portfile.cmake:28]: This function has been deprecated too. It should be `vcpkg_cmake_install()`

File: ports/urho3d/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+if(EXISTS ${CURRENT_PACKAGES_DIR}/share/Urho3D/CMake/Modules)
+    vcpkg_fixup_cmake_targets(CONFIG_PATH share/Urho3D/CMake/Modules)
+endif()
+if(EXISTS ${CURRENT_PACKAGES_DIR}/share/CMake/Modules)
+    vcpkg_fixup_cmake_targets(CONFIG_PATH share/CMake/Modules)
```
Comment (NancyLi1013) [inline ports/urho3d/portfile.cmake:33]: The same as above. It should be `vcpkg_cmake_config_fixup`.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by koprok** [comment]
@NancyLi1013 , done. I have made all changes that you have requested.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM now, thanks for adding this port. @koprok

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Pull request merged**